### PR TITLE
Ensure that post thumbnail is cached in latest post block.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## What?
Ensure that post thumbnails are correctly primed when using the last post block. 

## Why?
If this function is not called, it results in one query to post meta table. Calling `update_post_thumbnail_cache` primes all caches in one.

### CODE REVIEW

**Comment by getdave** [comment]
@spacedmonkey I've reached out to a few folks to see if we can get a review from a PHP-focused dev on this.

File: packages/block-library/src/latest-posts/index.php
Comment (hellofromtonya) [inline packages/block-library/src/latest-posts/index.php]: Guard against the key not being defined in the attributes:

```suggestion
	if ( isset( $attributes['displayFeaturedImage'] ) && $attributes['displayFeaturedImage'] ) {
```

Comment (spacedmonkey) [reply packages/block-library/src/latest-posts/index.php]: @hellofromtonya This is not done on this line 

```
if ( $attributes['displayFeaturedImage'] && has_post_thumbnail( $post ) ) {
```

See line 77. Should do this here as well?

Comment (hellofromtonya) [reply packages/block-library/src/latest-posts/index.php]: Yes. Why? There's nothing in this function that ensures that key exists in the attributes. These guards prevent unnecessary PHP Warning (in PHP >= 8.0) or Notice (< 8.0) https://3v4l.org/UNiKf.

Another way to go is to initialize the attributes before using, i.e. if not defined, set to `false`.

**Review by hellofromtonya — APPROVED** [review]
LGTM 👍

**Review by felixarntz — COMMENTED** [review]
@spacedmonkey Awesome work! Can you add a unit test, specifically around the cache bit for featured images? To ensure no additional request per featured image is fired.

**Comment by spacedmonkey** [comment]
@felixarntz I would love to, but I don't where I would do this. That test should really live in core. Do you have any recommendation on how I would add a test here.

**Comment by felixarntz** [comment]
> @felixarntz I would love to, but I don't where I would do this. That test should really live in core. Do you have any recommendation on how I would add a test here.

I think it would still work within Gutenberg's codebase, no? I think if we don't add tests here, we probably won't later add them to core either, since Gutenberg largely just gets merged into core over time.

You could add a file `class-block-library-latest-posts-test.php` to https://github.com/WordPress/gutenberg/tree/trunk/phpunit and implement a test using `parse_blocks` there (see https://github.com/WordPress/gutenberg/blob/trunk/phpunit/class-block-library-navigation-link-test.php#L96 as reference example for how GB code already tests other specific blocks).

**Comment by hellofromtonya** [comment]
>I think it would still work within Gutenberg's codebase, no? I think if we don't add tests here, we probably won't later add them to core either, since Gutenberg largely just gets merged into core over time.

I agree with @felixarntz. The PHPUnit tests are _also_ backported to Core along with the source code changes. But if there are no tests, then none are added to Core, meaning changes are backported without tests being added. Creating them now not only validates them during the R&D, stabilization, and maturity process in Gutenberg, but also ensures there are tests to backport once it lands in Core.

**Comment by spacedmonkey** [comment]
@felixarntz @hellofromtonya Point taken, unit tests added. Work noting, I don't like query counting tests, so I found another way of testing this.

**Comment by spacedmonkey** [comment]
@felixarntz @hellofromtonya Point taken, unit tests added. Worth noting, I don't like query counting tests, so I found another way of testing this.

**Comment by spacedmonkey** [comment]
@felixarntz @hellofromtonya Point taken, unit tests added. Worth noting, I don't like query counting tests, so I found another way of testing this.

File: phpunit/blocks/render-last-posts-test.php
Comment (costdev) [inline phpunit/blocks/render-last-posts-test.php]: Per [Using Assertions](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions), a test method with multiple assertions should pass a `$message` parameter to describe what failed:
> All PHPUnit assertions, as well as all WordPress custom assertions, allow for a `$message` parameter to be passed. This message will be displayed when the assertion fails and can help immensely when debugging a test. This parameter should always be used if more than one assertion is used in a test method.

File: packages/block-library/src/latest-posts/index.php
Comment (costdev) [inline packages/block-library/src/latest-posts/index.php:61]:
```suggestion
	if ( ! empty( $attributes['displayFeaturedImage'] ) ) {
```
This can be simplified to a single `! empty()` call.

Comment (spacedmonkey) [reply packages/block-library/src/latest-posts/index.php:61]: This is feedback from @hellofromtonya - https://github.com/WordPress/gutenberg/pull/40571#discussion_r858859870

File: phpunit/blocks/render-last-posts-test.php
Comment (peterwilsoncc) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
	public static function wpTearDownAfterClass() {
		wp_delete_post( self::$attachment_id, true );
	}
```

Possibly a comment this is deleted in the tear down as you need to delete the file from the uploads directory, which the base class tear down doesn't do.

File: phpunit/blocks/render-last-posts-test.php
Comment (peterwilsoncc) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
	 * @covers ::render_block_core_latest_posts
```

File: phpunit/blocks/render-last-posts-test.php
Comment (peterwilsoncc) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
	 * @covers ::render_block_core_latest_posts
```

File: phpunit/blocks/render-last-posts-test.php
Comment (peterwilsoncc) [inline phpunit/blocks/render-last-posts-test.php]: `/** @var */` docblocks would be dandy x 3

File: phpunit/blocks/render-last-posts-test.php
Comment (peterwilsoncc) [inline phpunit/blocks/render-last-posts-test.php:29]: Shared fixtures 😻

**Review by peterwilsoncc — COMMENTED** [review]
A couple of notes in the tests. 

It's a shame `update_post_thumbnail_cache()` requires a switch to `WP_Query` but I understand what's happening.

File: packages/block-library/src/latest-posts/index.php
Comment (costdev) [reply packages/block-library/src/latest-posts/index.php:61]: @hellofromtonya, what are your thoughts on this? Is there a case where `! empty()` would return undesired results?

File: phpunit/blocks/render-last-posts-test.php
Comment (spacedmonkey) [reply phpunit/blocks/render-last-posts-test.php]: Fixed.

File: packages/block-library/src/latest-posts/index.php
Comment (hellofromtonya) [reply packages/block-library/src/latest-posts/index.php:61]: Technically, yes, both patterns result in the same boolean state https://3v4l.org/q37iv. The design used made sure the key existed before the truthy check, whereas `empty()` hides away the intent of checking that the key actually exists.

But there's a missed opportunity here: What should the data type be for the `'displayFeaturedImage'` attribute? I'm assuming it's a true or false / turn it on or off state and not any non-falsey value. Right? 

What do I mean? I'm assuming this attribute expects a true or false to mean "yes turn this on" or "no turn it off". So it would not expect a non-empty array or string value (which PHP would juggle to a truthy state). Rather, it would expect a boolean or integer/string representation of a boolean. It should not expect any truthy or falsey value that PHP juggles into a boolean.

IMO This conditional needs to:
- ensure the attribute key exists AND
- its value is a `true` state: boolean `true`, string `'1'`, or integer `1`

What value and data type(s) will be in this specific attribute?

Comment (spacedmonkey) [reply packages/block-library/src/latest-posts/index.php:61]: I personally agree with Tonya on this one. I think it makes the code more readable and clear with the isset.

Comment (spacedmonkey) [reply packages/block-library/src/latest-posts/index.php:61]: That value should always be boolean and default to false. 

https://github.com/WordPress/gutenberg/blob/033a06f731c333f2c3fe066cc01c543614c34c30/packages/block-library/src/latest-posts/block.json#L60-L63

Comment (hellofromtonya) [reply packages/block-library/src/latest-posts/index.php:61]: If the attributes can be filtered, the value and data type could change. If that's the case, then my preference is this pattern:

```suggestion
isset( $attributes['displayFeaturedImage'] ) && true === $attributes['displayFeaturedImage']
```

Else, if the attributes are not filterable, then either `isset() && $the_variable` or `! empty()` patterns work the same. Then it's preference and one that should be documented in the handbook for consistency.

Comment (hellofromtonya) [reply packages/block-library/src/latest-posts/index.php:61]: That said, both patterns exist in Core and Gutenberg. IMO the discussion of which pattern to use likely belongs in a separate coding standard discussion / issue. Once there's consensus, then the pattern used here can be changed without effect or BC concerns.

File: phpunit/blocks/render-last-posts-test.php
Comment (anton-vlasenko) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
		$args      = $action->get_args();
```

File: phpunit/blocks/render-last-posts-test.php
Comment (anton-vlasenko) [inline phpunit/blocks/render-last-posts-test.php]: Nit pick: I understand this is "just" a test, but I believe it's better to use meaningful and pronounceable variable names.
```suggestion
		$action = new MockAction();
		add_filter( 'update_post_metadata_cache', array( $action, 'filter' ), 10, 2 );
```

File: phpunit/blocks/render-last-posts-test.php
Comment (anton-vlasenko) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
		$action = new MockAction();
		add_filter( 'update_post_metadata_cache', array( $action, 'filter' ), 10, 2 );
```

File: phpunit/blocks/render-last-posts-test.php
Comment (anton-vlasenko) [inline phpunit/blocks/render-last-posts-test.php]:
```suggestion
		$args      = $action->get_args();
```

File: phpunit/blocks/render-last-posts-test.php
Comment (anton-vlasenko) [inline phpunit/blocks/render-last-posts-test.php:68]: Would it be possible to add a message in case this test fails?
```suggestion
		$this->assertSameSets( self::$attachment_ids, $last_args[1], 'Attachment IDs must match' );
```

**Review by anton-vlasenko — CHANGES_REQUESTED** [review]
I have a few remarks about the test.
Thank you.

File: packages/block-library/src/latest-posts/index.php
Comment (costdev) [reply packages/block-library/src/latest-posts/index.php:61]: Agreed Tonya, about both the pattern discussion and the preferred pattern if the attributes can be filtered.

**Review by costdev — APPROVED** [review]

**Review by anton-vlasenko — APPROVED** [review]
LGTM.

**Comment by adamziel** [comment]
I cherry-picked this to wp/6.0 branch to be included in WordPress 6.0 RC2 later today a2be55c04f

**Comment by gziolo** [comment]
We will have to bring unit tests to WordPress core in a dedicated patch.

**Pull request merged**
